service/storage_service.cc

```diff
@@ -518,6 +518,12 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
     for (const auto& [id, rs]: t.transition_nodes) {
         co_await process_transition_node(id, rs);
     }
+    for (auto id : t.get_excluded_nodes()) {
+        locator::node* n = tmptr->get_topology().find_node(locator::host_id(id.uuid()));
+        if (n) {
+            n->set_excluded(true);
+        }
+    }
```

Comment:
So what happens after current operation completes and next one starts. We stop marking those nodes as excluded again. So it looks like we use locator as a way to pass current topology operation parameters somewhere. This somewhere looks to be the tablet load balancer. Can it get raft topology object directly instead?

Reply:
> So what happens after current operation completes and next one starts. We stop marking those nodes as excluded again. 

> Can it get raft topology object directly instead?

It's cleaner to have one mechanism for propagating topology snapshot, so I think existing users of token_metadata_ptr should keep using it and the long term solution is to move raft topology to locator topology, not propagating two kinds of topologies everywhere.

Reply:
> So what happens after current operation completes and next one starts. We stop marking those nodes as excluded again.

Yes. That's why I wanted excluded nodes to be permanent, but that's left for https://github.com/scylladb/scylladb/issues/16527

It shouldn't cause problems in the happy path, since nodes which are being removed/replaced will be removed from topology too when operation ends. Nodes which are marked as ignored will be unignored and that may block migrations. That could happen even before we started topology operation which ignored them, so not a new problem.

Reply:
If removenode fails and is aborted, nodes will be unexcluded too. This also falls into the same scenario and will be dealt with as part of https://github.com/scylladb/scylladb/issues/16527.

Reply:
> > Can it get raft topology object directly instead?
> 
> It's cleaner to have one mechanism for propagating topology snapshot, so I think existing users of token_metadata_ptr should keep using it and the long term solution is to move raft topology to locator topology, not propagating two kinds of topologies everywhere.

I also think that we should not have th same information in mutiple places, but in my view in the end token_metadata should be responsible for data placement only. Everything else should be in the topology.

Reply:
> I also think that we should not have th same information in mutiple places, but in my view in the end token_metadata should be responsible for data placement only. Everything else should be in the topology.

This patch is putting things to locator::topology, which is the right place. It's just that currently topology is carried by token_metadata_ptr, which is the "root" of the topology state. Maybe long term we should do some renaming there, and make it topology_ptr and have vnode part be a member of it.